caas/kubernetes/clientconfig/k8s.go

```diff
@@ -201,22 +201,31 @@ func credentialsFromConfig(config *clientcmdapi.Config, credentialName string) (
 			attrs["ClientCertificateData"] = string(user.ClientCertificateData)
 			hasCert = true
 		}
-		if len(user.ClientKeyData) > 0 {
+		hasClientKeyData := len(user.ClientKeyData) > 0
+		if hasClientKeyData {
 			attrs["ClientKeyData"] = string(user.ClientKeyData)
 		}
-
-		var authType cloud.AuthType
-		if user.Token != "" {
+		hasToken := user.Token != ""
+		if hasToken {
 			if user.Username != "" || user.Password != "" {
 				return cred, errors.NotValidf("AuthInfo: %q with both Token and User/Pass", name)
 			}
 			attrs["Token"] = user.Token
+		}
+
+		var authType cloud.AuthType
+		if hasClientKeyData {
+			// auth type used for aks for example.
+			authType = cloud.OAuth2AuthType
 			if hasCert {
 				authType = cloud.OAuth2WithCertAuthType
-			} else {
-				authType = cloud.OAuth2AuthType
+			}
+			if !hasToken {
+				// the Token is required.
+				return cred, errors.NotValidf("no Token found for %q with auth type %q", name, authType)
```

Comment:
"missing token for %v with auth type %v"

because NotValidf() adds " not valid" to the error message and it has to make sense with that

Reply:
yes, fixed!